source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

```diff
@@ -54,10 +173,32 @@ SPIFFEValidator::SPIFFEValidator(const Envoy::Ssl::CertificateValidationContextC
     }
   }
 
-  const auto size = message.trust_domains().size();
-  trust_bundle_stores_.reserve(size);
+  const auto n_trust_domains = message.trust_domains().size();
+
+  tls_->set([](Event::Dispatcher&) { return std::make_shared<ThreadLocalSpiffeState>(); });
+
+  // If a trust bundle map is provided, use that...
+  if (message.has_trust_bundles()) {
+    if (!message.trust_bundles().has_filename()) {
+      throw EnvoyException("SPIFFE Bundle DataSource requires a filename");
+    }
+    trust_bundle_file_name_ = message.trust_bundles().filename();
+    spiffe_data_ = loadTrustBundles();
+    if (!spiffe_data_) {
+      throw EnvoyException("Failed to load SPIFFE Bundle map");
+    }
+
+    updateSpiffeData(spiffe_data_);
+    initializeCertificateRefresh(context);
+    return;
+  }
+
```

Comment:
I think there should be two steps:
1. initialize the initial configuration that read from the data source by the `DataSource::read`.
2. set the file watching if necessary. tls_ should be initialized here if we require the watching.